### DISCUSSION TITLE
Adding drop shadow and smoothing to font command

### DISCUSSION THREAD

**Discussion by webbooo** [ImageMagick/ImageMagick#5032] [Help] [answered]
Hi all,

newbie here,  had an existing thread where I needed to select line items from a txt file and save each one as a separate png.

I had some great advice from @snibgo in a previous completed thread.  While this gave me a solution I realized the format of the font needed some formatting and am trying to add a shadow of sorts to the text or a way of smoothing it so it looks a bit more pleasing.

The below batch file has the command I'm using

set NUM=0

for /F "tokens=*" %%L in (sample.txt) do (
  setlocal enabledelayedexpansion
  %IMG7%magick -size 500x50 -gravity West -pointsize 40 -background pink -font courier -weight bold -style Oblique caption:"%%L" "%%L.png"
  set /A NUM+=1
)

The output looks like this
![Capture](https://user-images.githubusercontent.com/103241119/162557456-3fe8785c-ef2a-4d38-8b27-901cfb45a955.JPG)

I'm getting used to the different commands but havent been able to find a way to add a drop shadow (white) or to have it so the font is smooth and bold. Would like to have it so it slightly indented from the left also.

Thanks
Adam

**Comment by snibgo** [comment]
I suggest moving "setlocal enabledelayedexpansion" to the top of the script.

I use %IMG7% because that is where my IM installation resides. Unless you do the same, you should remove that.

For drop shadows, see https://legacy.imagemagick.org/Usage/blur/#shadow.

To add a drop shadow, you need to start with an opaque object (such as text) on a transparent background. But your image is entirely opaque, so the shadow would be of the entire image, not just the text. You can write the text over a transparent background, then create the shadow, then add the pink by flattening the shadow and text against that colour.
```
setlocal enabledelayedexpansion

for /F "tokens=*" %%L in (sample.txt) do (
  magick ^
    -size 300x300 -gravity SouthWest -pointsize 40 -background None ^
    -font Arial-Bold ^
    caption:"%%L" ^
    -repage +30+0 ^
    ^( +clone ^
       -background Navy  -shadow 80x3+5+5 ^
    ^) +swap ^
    -background Pink -layers Flatten  +repage ^
    "%%L.png"
)
```
I have used a bold font. "magick -list font" lists the available fonts.

"-repage +30+0" gives an offset of 30 pixels to the right.

We no longer need variable "NUM".

**Reply by webbooo** [reply]
thank you mate,

I'm learning a lot from you, appreciate it. I noticed you removed the gravity command, have played around with repage and that works really well.

Is it possible to keep the image at this size but to only have the background relative to the size of the text or at a set width height also? example below.  Would something like that require much change?

![Capture](https://user-images.githubusercontent.com/103241119/162598038-b5fd31f6-cdd9-47f9-aef7-c1d98167f5fe.JPG)

setlocal enabledelayedexpansion

for /F "tokens=*" %%L in (sample.txt) do (
  magick ^
    -size 300x300 -gravity SouthWest -pointsize 20 -background None ^
    -font Arial-Bold ^
    caption:"%%L" ^
    -repage +10-270 ^
    ^( +clone ^
       -background White  -shadow 80x3+5+5 ^
    ^) +swap ^
    -background RoyalBlue2 -layers Flatten  +repage ^
    "%%L.png"
)

cheers
Adam

**Reply by webbooo** [reply]
This program seem to be a little more complicated than I thought, for some reason the output files are 511x511 in size and not 500x500 which is bizarre,  Not sure why thats happening, they are also 64 bit depth when I thought theyd be 32.  When I overlayed the files on another png they lost their quality. Golly Gosh..

**Comment by fmw42** [comment]
If you just want a transparent background to your text, you can do

`magick -size 300x300 -gravity center -pointsize 20 -background none -font Arial-Bold caption:"some text" result`

If that is not what you want, please clarify in more detail

**Comment by webbooo** [comment]
Hi thanks,

I'd like only the text itself to have a solid background that's relative to the size of the txt.  The rest of the image can be transparent.

Hopefully this makes sense.

Thanks

**Reply by fmw42** [reply]
Thats what you have above from snibgo. So, I still confused. You have not made your request clear.  Perhaps make an example image in GIMP to show us what you want different.

**Reply by webbooo** [reply]
sorry, I thought it was relatively clear as I've also included it in the image above.  Yes I've received some great support from @snibgo who has been very helpful.

** Just to be clear this is what the script currently outputs which has worked very well.

![Capture1](https://user-images.githubusercontent.com/103241119/162604684-c651d15a-df5e-431b-bfba-c4021b1b22f1.JPG)

** This is what I queried if it is possible to achieve.  In this example you can see it only the text the has the solid background.  The rest of the image is transparent as I'd still like the png itself to stay at 300x300.  In the sample above the entire image has a solid background and I'm trying to change this so only the text itself has the solid background while still keeping the 300x300.

![Capture](https://user-images.githubusercontent.com/103241119/162604608-e3b3fb61-f829-46be-b496-f4b1398c1acb.JPG)

Thanks
Adam

**Comment by snibgo** [comment]
I assume the chequerboard in the required output represents transparent pixels.

You can create two images. The first is a 300x300 transparent image. The second is the text. Then composite the two images.
```
for /F "tokens=*" %%L in (sample.txt) do (
  magick ^
    -size 300x300 xc:None ^
    ^( -gravity Center -pointsize 40 -background None ^
       -font Arial-Bold ^
       caption:"%%L" ^
       ^( +clone ^
          -background Navy -shadow 80x3+5+5 ^
       ^) ^
       +swap ^
       -background Pink -layers Flatten +repage ^
       -trim +repage ^
    ^) ^
    -gravity NorthWest ^
    -compose Over -composite ^
    "%%L.png"
)
```

>  for some reason the output files are 511x511 in size and not 500x500 which is bizarre, Not sure why thats happening, ...

Did you read the link I gave you? It says:
> The problem with shadow is that a blurry shadow extends in all directions. To compensate the "-shadow" operator enlarges the actual original image by adding a border 2 times the size of the blur 'sigma' value given.

> ...they are also 64 bit depth when I thought theyd be 32.

PNG files are 8 or 16 bits per channel. You need 4 channels (RGB and alpha). By default, IM uses the smallest of 8 or 16 that is required for the data. If you want to force 8 bits, use "-depth 8" before writing the file.

> When I overlayed the files on another png they lost their quality. Golly Gosh..

I don't understand that.

**Reply by webbooo** [reply]
Thanks,  this worked but I'm finding when I'm using the png that has been created to over lay with another the output looks like the sample here. So I thought there must be an issue with the bit depth or color profile.  I have a different script I'm using to over lay the png with the persons photo so this part isnt dont in imagemagick.  I have about 2500 of these I'll need to do by the end and as you can see the end result is that the person will have their name placed over their picture.

![1](https://user-images.githubusercontent.com/103241119/162642458-fc5aff35-7c07-4515-a2b4-114a221edb02.png)

Thanks
Adam

**Reply by webbooo** [reply]
Just to be clear.  The script you helped produced is working in that its producing the required transparent png file from text.  Issue I have is when I overlay that png onto another (persons photo).

Thanks
Adam

**Comment by fmw42** [comment]
> I have a different script I'm using to over lay the png with the persons photo

We cannot help if you do not post your code!

**Reply by webbooo** [reply]
the code is above, why you always quick to jump on people?  I'm just reposting what is above.  This code works as I said again in that it produces the required png (has transparent png with text and the text has a solid border.  The issue I have is when I go and use the png that the script produces the text and shadow become all distorted. as per sample above.

```
setlocal enabledelayedexpansion

for /F "tokens=*" %%L in (sample.txt) do (
  magick ^
    -size 300x300 xc:None ^
    ^( -gravity Center -pointsize 40 -background None ^
       -font Arial-Bold ^
       caption:"%%L" ^
       ^( +clone ^
          -background Navy -shadow 80x3+5+5 ^
       ^) ^
       +swap ^
       -background Pink -layers Flatten +repage ^
       -trim +repage ^
    ^) ^
    -gravity NorthWest ^
    -compose Over -composite ^
    "%%L.png"
)
```

**Reply by fmw42** [reply]
Try adding `-compose over` before -layers flatten to reset the compose setting.  Does that work?

**Reply by webbooo** [reply]
no this didnt work. Thanks for the suggestion.

I think the issue is the resolution of the png, ie how many pixels per inch imagemagick is using on the png files that the script is creating. Is there a way to set this?  72 pixels per inch would be great.

**Reply by fmw42** [reply]
Resolution or density will not affect the digital image only printing of it. So something else is wrong.  Sorry, I am not a Windows users, so cannot really test your code easily without converting to bash unix.

**Reply by fmw42** [reply]
I do not see where you overlay the result of that code on your background image of the person.  What format is that image?  Is it RGB or CMYK?  Please post the full code including the overlay onto the background person image

**Reply by webbooo** [reply]
ok no worries, thanks for your help anyway.  I thought resolution or colour profile etc could be the issue but may be wrong.  If I recreate the png that the script creates in Photoshop (with higher pixel count) it doesn't become an issue.

**Reply by webbooo** [reply]
something you may know, when setting a color, instead of chosing pink or green etc is there a command that can make it random? 

so instead of 

`-background Black -compose Over -layers Flatten +repage ^`

something like

`-background Ranom -compose Over -layers Flatten +repage ^`

**Comment by snibgo** [comment, marked as answer]
>  I have a different script I'm using to over lay the png with the persons photo so this part isnt dont in imagemagick.

If you can share an input image (a person's photo) that gives the problem, we can check if there is something weird about it. For example, if it is encoded as CMYK, then compositing a sRGB image over it won't work; they should both be converted to sRGB first.

I don't think CMYK/sRGB is the problem, that's just an example.

On random colours: sadly, "-background" and "xc:sRGB(...)" can't take fx expressions. We can make an entire image a constant random colour like this:
```
magick -size 100x100 xc:White ^
  -channel R -evaluate Multiply %[fx:rand()] ^
  -channel G -evaluate Multiply %[fx:rand()] ^
  -channel B -evaluate Multiply %[fx:rand()] ^
  +channel ^
  out.png
```
So instead of "-background Pink", we can start with an image of random colour:
```
for /F "tokens=*" %%L in (sample.txt) do (
  %FXNEW%magick ^
    -size 300x300 xc:None ^
    ^( ^
        ^( xc:White ^
           -channel R -evaluate Multiply "%%[fx:rand()]" ^
           -channel G -evaluate Multiply "%%[fx:rand()]" ^
           -channel B -evaluate Multiply "%%[fx:rand()]" ^
           +channel ^
        ^) ^
        ^( ^
           -gravity Center -pointsize 40 -background None ^
           -font Arial-Bold ^
           caption:"%%L" ^
       ^) ^
       ^( +clone ^
          -background Navy -shadow 80x3+5+5 ^
       ^) ^
       +swap ^
       -background None -layers Flatten +repage ^
       -trim +repage ^
    ^) ^
    -gravity NorthWest ^
    -compose Over -composite ^
    "%%L.png"
)
```
Some colours will be dark, so instead of "%%[fx:rand()]" you could have "%%[fx:0.6+rand()*0.4]" or whatever, to ensure you get light colours.

**Reply by webbooo** [reply]
Thanks, this works really well with the randomizing of the background colours.  Very Cool.

Here is a sample picture
 
![Mary Clark](https://user-images.githubusercontent.com/103241119/162644788-2e9e339f-46a9-4805-a6bd-f6f0999ef23b.png)

When I overlay the png that your script creates on this png it ends up something like this

![2](https://user-images.githubusercontent.com/103241119/162644827-f87c4d0a-1134-4c2d-aafb-e318af69a550.png)

When I recreate the png from your script in photoshop (copy and paste into a new image) and then overlay it I get this.

![1](https://user-images.githubusercontent.com/103241119/162644896-98e9c2ca-6491-41ac-a59b-83b937d53a1c.png)

**Reply by fmw42** [reply]
Sorry, but you did not provide your overlay code.

**Reply by webbooo** [reply]
understand but as Ive said I'm not using imagemagick for the overlay part so didnt think there is any point in posting it here as its not imagemagick code? - its javascript

If there is a separate way of doing the overlay in imagemagick then perhaps I should consider that?

In the end I have 2 folders with about 2500 photos.  One folder has the png of the users name (thanks to this thread) and the other folder has the png of the users photo.  The last part is for me to combine the 2.

Thanks

**Comment by fmw42** [comment]
`magick background_image overlay_image -geometry +X+Y -gravity some_gravity -compose -over -composite result_image`

But you can fold that into your script as well.

**Reply by webbooo** [reply]
where do I define the source png and the overlay png and how can I get it to run through the 2500 of them?

**Reply by fmw42** [reply]
`magick background_image.png overlay_image.png -geometry +X+Y -gravity some_gravity -compose -over -composite result_image.png`

**Reply by webbooo** [reply]
sorry this is a little confusing, as I would need to update this 2500 times for each overlay?

lets say I have the png with the names in c:\usernames\

in here there are 2500 png files of the user names eg

Jo James text.png
Sam Rigg text.png
Donald Pump text.png

etc etc

In a second folder, lets say c:\userphotos\ I have all the user profile pictures

in here there are 2500 png files of the user photos eg

Jo James Photo.png
Sam Rigg Photo.png
Donald Pump Photo.png

The last part I'm facing is trying to overlay Jo James text.png onto Jo James Photo.png,  I have been trying a javascript for this but have been having the trouble mentioned above and how the image comes out.

Using your suggestion this would mean I'd have to go through each one and manually change the criteria would it? 

ie 

`magick Jo James Photo.png Jo James text.png -geometry +X+Y -gravity some_gravity -compose -over -composite Jo James Final.png`

I can change the names to the png files to whatever would make it easier, its just finding a script that will go through the 2500 without having to change the script for each user etc.

Thanks

**Reply by fmw42** [reply]
You will have to wait for snibgo to help you fold that into your other script. Sorry, I do not script Windows. I posted that code, so that you could just try doing it in ImageMagick and see if you get the same issue.

**Reply by webbooo** [reply]
ahh I see, sorry.  I will try that and see.

Thanks for the suggestion

**Reply by webbooo** [reply]
i got this error when trying your suggestion

magick: invalid argument for option '-geometry' '+X+Y' at CLI arg 3 @ error/operation.c/CLISimpleOperatorImage/2518.

**Reply by fmw42** [reply]
You have to provide an X and Y value for the offset from the nominal gravity location. If you want the nominal, then just use +0+0.

**Reply by fmw42** [reply]
Please refer to the documentation for description of settings.

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli
https://legacy.imagemagick.org/discourse-server/index.php

**Reply by webbooo** [reply]
lol completely overlooked that, thanks for the links

**Comment by snibgo** [comment]
It would be easier if the names were the same in the two directories, eg "Jo James.png" in both "c:\usernames\" and "c:\userphotos\".

Then a Windows BAT script can loop through file in "c:\userphotos\". For each one that it finds there, it runs the command above, using variable for the filenames. Like this (untested):
```
for %%F in (c:\userphotos\*.png) do magick "%%F" "c:\usernames\%%~nF.png" -gravity NorthWest -compose Over -composite "c:\outdir\%%~nF.png"
```
If the files are named "Jo James Photo.png" and "Jo James Text.png", it can still be done, but it's more complex.

**Reply by webbooo** [reply]
I see, windows doesn't let you have more than one file name with the same name and file type in a folder, it needs something appended to it or a different file type.

**Comment by snibgo** [comment]
> I see, windows doesn't let you have more than one file name with the same name and file type in a folder, it needs something appended to it or a different file type.

That is true, but is not a problem. We can't have two files both named "Jo James Text.png" in the same directory. But we can have files with the same name if they are in different directories. If they have identical names in the two input directories (and the output directory), then matching them up is easy.

**Reply by webbooo** [reply]
yes, they can be in different directories with the same name.  I can place the files where ever is needed
